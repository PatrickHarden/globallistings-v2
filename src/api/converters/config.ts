import { mapObjIndexed } from 'ramda';
import { Config } from '../../types/config';
import flattenObject from '../../utils/flattenObj';

const replaceIntlChars = mapObjIndexed((_, key, obj) =>
  obj[key].replace(/%\(/g, '{').replace(/\)s/g, '}')
);

const convertToIntl = (messages: object) => replaceIntlChars(messages);

// TODO this should leave outside API folder as this is not an API layer anymore and just a transformation layer
export const convert = (config: Config): Config => ({
  ...config,
  i18n: convertToIntl(flattenObject(config.i18n)),
});
