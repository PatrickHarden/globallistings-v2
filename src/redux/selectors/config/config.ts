import { createSelector } from 'reselect';
import { configSelector, configMetaSelector } from '../state';

export const configStatusSelector = createSelector(
  configMetaSelector,
  config => config.status
);

export const getFilters = createSelector(
  configSelector,
  config => config.filters
);

export const siteTypeSelector = createSelector(
  configSelector,
  config => config.siteType
);
export const searchModeSelector = createSelector(
  configSelector,
  config => config.searchMode
);

// SearchConfig
export const searchConfigSelector = createSelector(
  configSelector,
  config => config.searchConfig
);

export const searchBiasSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchBias
);

export const searchRadiusRestrictionSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.biasRadius
);

export const searchCountryRestrictionSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.restrictToCountry
);

export const polygonSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.polygons
);

export const searchTypesRestrictionSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchPlaceTypes
);

export const assumeZipCodeSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.allowAssumeZipCode
);

export const searchPathSortAlphabeticalSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchPathSortAlphabetical
);

export const searchPathSelectorSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchPathSelector
);

export const propertyIdPathSelectorSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.propertyIdSearchPathSelector
);

export const hideBuyButtonSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.hideSearchToBuy
);

export const hideLetButtonSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.hideSearchToLet
);

export const getSearchHeader = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchHeader
);
export const searchResultsPageSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchResultsPage
);

export const searchSuggestDebounceSelector = createSelector(
  searchConfigSelector,
  searchConfig => searchConfig.searchSuggestDebounce
);

export const defaultSearchLocationNameSelector = createSelector(
  configSelector,
  config => config.searchLocationName
);

export const paramsSelector = createSelector(
  configSelector,
  config => config.params
);

export const restrictionsSelector = createSelector(
  configSelector,
  config => ({
    country: config.searchConfig.restrictToCountry,
  })
);

export const otherConfigsSelector = createSelector(
  configSelector,
  config => (config.otherConfigs ? config.otherConfigs : [])
);

export const currentConfigSelector = createSelector(
  configSelector,
  config => config.currentConfig
);
