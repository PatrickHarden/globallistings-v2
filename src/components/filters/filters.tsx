import React, { FC } from 'react';
import styled from 'styled-components';
import { FilterInfo } from '../../redux/selectors/config/filters';
import { RangeFilter } from './range-filter';
import { SelectFilter } from './select-filter';
import { FilterModifier } from '../../types/config';

const Wrapper = styled.div``;

interface Props {
  readonly displayLabels: boolean;
  readonly filters: FilterInfo[];
  readonly handleChange: (key: string) => (val: string, modifier: FilterModifier) => void;
  readonly inline: boolean;
  readonly placement?: string;
  readonly className?: string;
}

const filterToKey = (filter: FilterInfo) => `${filter.type}-${filter.name}`;

export const Filters: FC<Props> = ({
  className,
  displayLabels,
  filters,
  handleChange,
  inline,
  placement,
}) => (
  <Wrapper className={className}>
    {filters
      .filter(filter => (placement ? filter.placement === placement : true))
      .map(filter => {
        switch (filter.type) {
          case 'select':
            return (
              <SelectFilter
                displayLabel={displayLabels}
                handleSelect={handleChange(filter.name)}
                inline={inline}
                filter={filter}
                key={filterToKey(filter)}
              />
            );
          case 'range':
            return (
              <RangeFilter
                displayLabel={displayLabels}
                filter={filter}
                key={filterToKey(filter)}
                handleChange={handleChange}
                inline={inline}
              />
            );
          default:
            console.error(`Unsupported filter type: ${filter.type}`);
            return null;
        }
      })}
  </Wrapper>
);
