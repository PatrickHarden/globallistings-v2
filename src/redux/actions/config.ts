import axios from 'axios';
import {
  createRequestPending,
  createRequestFailure,
  createRequestSuccess,
} from '../../types/common/action-creators-singular';
import { Config } from '../../types/config';
import { State } from '../../types/state';
import { configEndpointForCurrentUsageTypeSelector } from '../selectors/plp';

export const createFetchConfigRequest = createRequestPending('FETCH_CONFIG_REQUEST');
export const createFetchConfigFailure = createRequestFailure('FETCH_CONFIG_FAILURE');
export const createFetchConfigSuccess = createRequestSuccess<Config, 'FETCH_CONFIG_SUCCESS'>(
  'FETCH_CONFIG_SUCCESS'
);

// This will be used in the future to switch between multiple configs
export const fetchConfig = () => (dispatch: Function, getState: () => State) => {
  const state = getState();
  const configEndpoint = configEndpointForCurrentUsageTypeSelector(state);

  dispatch(createFetchConfigRequest());
  return axios
    .get(configEndpoint)
    .then(res => {
      dispatch(createFetchConfigSuccess(res.data));
    })
    .catch(ex => {
      dispatch(createFetchConfigFailure(ex));
    });
};
