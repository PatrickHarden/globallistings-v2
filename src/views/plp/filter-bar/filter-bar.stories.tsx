import React from 'react';
import { storiesOf } from '@storybook/react';
import { FilterBar } from './filter-bar';
import { FilterType, FilterPlacement, FilterModifier } from '../../../types/config';
import { RangeFilterInfo, SelectFilterInfo } from '../../../redux/selectors/config/filters';

const rangeFilter: RangeFilterInfo = {
  name: 'Dynamic.LettingCharge',
  placement: FilterPlacement.primary,
  type: FilterType.range,
  label: 'Rent',
  min: { items: [] },
  max: { items: [] },
  modifier: FilterModifier.query,
};

const selectFilter: SelectFilterInfo = {
  name: 'radius',
  placement: FilterPlacement.primary,
  type: FilterType.select,
  label: 'Radius',
  items: [],
  modifier: FilterModifier.query,
};

const filterChange = (key: string) => (c: string) => console.log(`${c} is selected`);
const onLocationChange = (c: string) => console.log(`${c} is selected`);

storiesOf('FilterBar', module)
  .add('Without Filters', () => (
    <FilterBar
      filters={[]}
      onFilterChange={filterChange}
      onLocationChange={onLocationChange}
      suggestions={[]}
      changePlaceSearch={filterChange}
    />
  ))
  .add('With One Range filter', () => (
    <FilterBar
      filters={[rangeFilter]}
      onFilterChange={filterChange}
      onLocationChange={onLocationChange}
      suggestions={[]}
      changePlaceSearch={filterChange}
    />
  ))
  .add('With Range and Select filter', () => (
    <FilterBar
      filters={[rangeFilter, selectFilter]}
      onFilterChange={filterChange}
      onLocationChange={onLocationChange}
      suggestions={[]}
      changePlaceSearch={filterChange}
    />
  ));
