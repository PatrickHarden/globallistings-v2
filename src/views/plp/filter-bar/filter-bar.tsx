import * as React from 'react';
import styled from 'styled-components';
import MapSearch from '../../../components/map-search/map-search';
import { Filters } from '../../../components/filters/filters';
import { FilterInfo } from '../../../redux/selectors/config/filters';
import { AutocompleteSuggestion } from '../../../redux/selectors/search';
import { FilterModifier } from '../../../types/config';

const Container = styled.div`
  border-bottom: ${props => props.theme.borders.fine};
  display: flex;
  flex-direction: row;
  align-items: center;
`;

const StyledFilters = styled(Filters)`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  align-items: center;
`;

const MapSearchWrapper = styled.div`
  flex: 2;
`;

const LocationSearchContainer = styled.div`
  display: flex;
  flex: 2;
`;

interface Props {
  readonly filters: FilterInfo[];
  readonly changePlaceSearch: (searchTerm: string) => void;
  readonly onFilterChange: (key: string) => (val: string, modifier: FilterModifier) => void;
  readonly onLocationChange: (value: string) => void;
  readonly suggestions: AutocompleteSuggestion[];
  readonly initialSearchLocation?: string;
}

export const FilterBar: React.SFC<Props> = ({
  changePlaceSearch,
  initialSearchLocation,
  filters,
  onFilterChange,
  onLocationChange,
  suggestions,
}) => (
  <Container>
    <LocationSearchContainer>
      <MapSearchWrapper>
        <MapSearch
          initialSearchLocation={initialSearchLocation}
          handleSelect={onLocationChange}
          handleInput={changePlaceSearch}
          suggestions={suggestions}
        />
      </MapSearchWrapper>
      <StyledFilters
        displayLabels={false}
        filters={filters}
        handleChange={onFilterChange}
        inline={true}
        placement={'lm_locationFilter'}
      />
    </LocationSearchContainer>

    <StyledFilters
      displayLabels={false}
      filters={filters}
      handleChange={onFilterChange}
      inline={true}
      placement={'lm_primaryFilter'}
    />

    <StyledFilters
      displayLabels={false}
      filters={filters}
      handleChange={onFilterChange}
      inline={true}
      placement={'lm_secondaryFilter'}
    />

    <StyledFilters
      displayLabels={false}
      filters={filters}
      handleChange={onFilterChange}
      inline={true}
      placement={'lm_tertiaryFilters'}
    />
  </Container>
);
