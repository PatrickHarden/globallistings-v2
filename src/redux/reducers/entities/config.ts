import { convert } from '../../../api/converters/config';
import { createFetchConfigSuccess } from '../../actions/config';
import { ActionType } from 'typesafe-actions';
import { DeepPartial } from 'redux';
import { Config } from '../../../types/config';

export type Actions = ActionType<typeof createFetchConfigSuccess>;

export default (state: DeepPartial<Config> = {}, action: Actions): Config | {} => {
  switch (action.type) {
    case 'FETCH_CONFIG_SUCCESS':
      return convert(action.payload);
    default:
      return state;
  }
};
