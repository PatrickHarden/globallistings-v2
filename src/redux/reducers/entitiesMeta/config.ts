import * as actions from '../../actions/config';
import { ActionType } from 'typesafe-actions';
import { handleError, handleSuccess, handlePending } from './shared/meta-reducers';

export type Actions = ActionType<typeof actions>;

export default (state = {}, action: Actions) => {
  switch (action.type) {
    case 'FETCH_CONFIG_REQUEST':
      return handlePending();
    case 'FETCH_CONFIG_SUCCESS':
      return handleSuccess(action);
    case 'FETCH_CONFIG_FAILURE':
      return handleError(action);
    default:
      return state;
  }
};
