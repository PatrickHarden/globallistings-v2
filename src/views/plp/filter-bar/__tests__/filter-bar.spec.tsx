import React from 'react';
import { shallow } from 'enzyme';

import { FilterType, FilterPlacement, FilterModifier } from '../../../../types/config';
import { FilterBar } from '../filter-bar';
import { FilterInfo } from '../../../../redux/selectors/config/filters';

describe(FilterBar, () => {
  const filters: FilterInfo[] = [
    {
      name,
      label: 'select filter',
      type: FilterType.select,
      items: [{ value: 'foo', label: 'bar', default: false }],
      placement: FilterPlacement.primary,
      modifier: FilterModifier.query,
    },
    {
      name,
      label: 'Ramge filter',
      type: FilterType.range,
      min: { items: [] },
      max: { items: [] },
      placement: FilterPlacement.secondary,
      modifier: FilterModifier.query,
    },
  ];
  const props = {
    filters,
    changePlaceSearch: (searchTerm: string) => searchTerm,
    onLocationChange: (searchTerm: string) => searchTerm,
    onFilterChange: (key: string) => (val: string, modifier: FilterModifier) => ({ val, modifier }),
    suggestions: [{ label: 'London', value: 'abc123', default: false }],
    initialSearchLocation: 'London',
  };

  it('should render correctly', () => {
    expect(shallow(<FilterBar {...props} />)).toMatchSnapshot();
  });
});
