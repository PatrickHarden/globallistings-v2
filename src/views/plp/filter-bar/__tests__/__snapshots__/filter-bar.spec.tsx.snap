// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterBar should render correctly 1`] = `
<styled.div>
  <styled.div>
    <styled.div>
      <MapSearch
        handleInput={[Function]}
        handleSelect={[Function]}
        initialSearchLocation="London"
        suggestions={
          Array [
            Object {
              "default": false,
              "label": "London",
              "value": "abc123",
            },
          ]
        }
      />
    </styled.div>
    <Styled(Filters)
      displayLabels={false}
      filters={
        Array [
          Object {
            "items": Array [
              Object {
                "default": false,
                "label": "bar",
                "value": "foo",
              },
            ],
            "label": "select filter",
            "modifier": "query",
            "name": "nodejs",
            "placement": "lm_primaryFilter",
            "type": "select",
          },
          Object {
            "label": "Ramge filter",
            "max": Object {
              "items": Array [],
            },
            "min": Object {
              "items": Array [],
            },
            "modifier": "query",
            "name": "nodejs",
            "placement": "lm_secondaryFilter",
            "type": "range",
          },
        ]
      }
      handleChange={[Function]}
      inline={true}
      placement="lm_locationFilter"
    />
  </styled.div>
  <Styled(Filters)
    displayLabels={false}
    filters={
      Array [
        Object {
          "items": Array [
            Object {
              "default": false,
              "label": "bar",
              "value": "foo",
            },
          ],
          "label": "select filter",
          "modifier": "query",
          "name": "nodejs",
          "placement": "lm_primaryFilter",
          "type": "select",
        },
        Object {
          "label": "Ramge filter",
          "max": Object {
            "items": Array [],
          },
          "min": Object {
            "items": Array [],
          },
          "modifier": "query",
          "name": "nodejs",
          "placement": "lm_secondaryFilter",
          "type": "range",
        },
      ]
    }
    handleChange={[Function]}
    inline={true}
    placement="lm_primaryFilter"
  />
  <Styled(Filters)
    displayLabels={false}
    filters={
      Array [
        Object {
          "items": Array [
            Object {
              "default": false,
              "label": "bar",
              "value": "foo",
            },
          ],
          "label": "select filter",
          "modifier": "query",
          "name": "nodejs",
          "placement": "lm_primaryFilter",
          "type": "select",
        },
        Object {
          "label": "Ramge filter",
          "max": Object {
            "items": Array [],
          },
          "min": Object {
            "items": Array [],
          },
          "modifier": "query",
          "name": "nodejs",
          "placement": "lm_secondaryFilter",
          "type": "range",
        },
      ]
    }
    handleChange={[Function]}
    inline={true}
    placement="lm_secondaryFilter"
  />
  <Styled(Filters)
    displayLabels={false}
    filters={
      Array [
        Object {
          "items": Array [
            Object {
              "default": false,
              "label": "bar",
              "value": "foo",
            },
          ],
          "label": "select filter",
          "modifier": "query",
          "name": "nodejs",
          "placement": "lm_primaryFilter",
          "type": "select",
        },
        Object {
          "label": "Ramge filter",
          "max": Object {
            "items": Array [],
          },
          "min": Object {
            "items": Array [],
          },
          "modifier": "query",
          "name": "nodejs",
          "placement": "lm_secondaryFilter",
          "type": "range",
        },
      ]
    }
    handleChange={[Function]}
    inline={true}
    placement="lm_tertiaryFilters"
  />
</styled.div>
`;
